server/types/types.go

```diff
@@ -440,6 +440,10 @@ func (err BadRequestErr) Error() string {
 
 // IsBadRequest returns true if err is a BadRequestErr.
 func IsBadRequest(err error) bool {
-	_, ok := err.(BadRequestErr)
+	errors, ok := err.(ast.Errors)
```

Comment:
I don't think we should overload this function to check for ast.Errors. It's a badly named error type that's only used for JSON patch-related errors at the moment.

The `v1QueryGet` and `v1QueryPost` methods on the server currently check the error returned by `execQuery`. I think what's currently happening is that parse errors are returned by the `rego` package as `rego.Errors`. One option would be to update the callers of `execQuery` to parse the query themselves and then provide a parsed query to `execQuery`. This would allow the callers to return 400 in the right circumstance.

Reply:
In order to pass a parsed query to execQuery the entire call path needs to change to take a different argument since Eval() would parse and compile the query again. Thoughts?

Reply:
`rego.Eval` can take a `rego.ParsedQuery` option instead of `rego.Query`.

Reply:
ok, I will add a new argument to the function.

Reply:
Right. It should be possible to:

1. Parse the query in the caller (e.g., in v1QueryGet) and pass the query to the execQuery function
2. Inside execQuery, use rego.ParsedQuery instead of rego.Query.